plugin/src/main/java/org/opennms/timeseries/cortex/ResultMapper.java

```diff
@@ -62,31 +62,45 @@ public static Sample toSample(final Metric metric, final JSONArray jarray) {
     }
 
     public static <T> Metric toMetricFromMap(Map<String, T> tags) {
-        final Set<Tag> intrinsicTags = new LinkedHashSet<>();
-        final Set<Tag> metaTags = new LinkedHashSet<>();
-        final Set<Tag> externalTags = new LinkedHashSet<>();
+        ImmutableMetric.MetricBuilder metric = ImmutableMetric.builder();
+
         for (Map.Entry<String, T> entry : tags.entrySet()) {
             final String labelName = entry.getKey();
             final String labelValue = entry.getValue().toString();
 
             if (METRIC_NAME_LABEL.equals(labelName)) {
-                intrinsicTags.add(new ImmutableTag(IntrinsicTagNames.name, labelValue));
-                continue;
-            }
-
-            final Tag tag = new ImmutableTag(labelName, labelValue);
-            if (INTRINSIC_TAG_NAMES.contains(labelName)) {
-                intrinsicTags.add(tag);
+                metric.intrinsicTag(IntrinsicTagNames.name, labelValue);
             } else if (labelName.startsWith("_ext")) {
-                externalTags.add(tag);
-            } else {
-                metaTags.add(tag);
+                metric.externalTag(externalLabelToTag(labelValue));
+            } else if (INTRINSIC_TAG_NAMES.contains(labelName)) {
+                metric.intrinsicTag(labelName, labelValue);
+            } else  {
+                metric.metaTag(labelName, labelValue);
             }
         }
-        return new ImmutableMetric(intrinsicTags, metaTags, externalTags);
+        return metric.build();
     }
 
     static <T> Stream<T> iteratorToFiniteStream(final Iterator<T> iterator) {
         return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, 0), false);
     }
+
+    static PrometheusTypes.Label.Builder externalTagToLabel(final Tag tag) {
+        // _ext_%hash% = %key%=%value%
+        // %hash% is a hash of the value - the key contains special characters we want to preserve, so we place it in the value instead
+        String name = "_ext_" + tag.getValue().hashCode();
```

Comment:
Let's use the hash of the value computed bellow (`tag.getKey() + "=" + tag.getValue()`) as opposed to the initial tag value.

Reply:
done.